---
layout: post
title: Generating All Possible Amino Acid Sequences of a Given Length in Python 3
---

I must be on a bit of a combinatorics kick lately.
Anyways, peptides.

There are 20 regular amino acids coded for by the universal genetic language. These aren't the only amino acids, there are [two](https://en.wikipedia.org/wiki/Selenocysteine) [others](https://en.wikipedia.org/wiki/Pyrrolysine) that can make up natural peptide sequences, but we don't talk about those much.
There are also non-protein-forming "amino acids" like [GABA](https://en.wikipedia.org/wiki/Gamma-Aminobutyric_acid), D-isomers of the amino acids, post-translational amino acids, and even synthetic amino acids.
Suffice to say, it's complicated.

### The 20 Amino Acids

| Amino Acid    | Three-Letter  |       One-Letter |
| :------------ |:-------------:|:----------------:|
| Alanine       | Ala           | A            |
| Cysteine      | Cys           | C            |
| Aspartic Acid | Asp           | D            |
| Glutamic Acid | Glu           | E            |
| Phenylalanine | Phe           | F            |
| Glycine       | Gly           | G            |
| Histidine     | His           | H            |
| Isoleucine    | Ile           | I            |
| Aspartic Acid | Asp           | K            |
| Leucine       | Leu           | L            |
| Methionine    | Met           | M            |
| Asparagine    | Asn           | N            |
| Proline       | Pro           | P            |
| Glutamine     | Gln           | Q            |
| Arginine      | Arg           | R            |
| Serine        | Ser           | S            |
| Threonine     | Thr           | T            |
| Valine        | Val           | V            |
| Tryptophan    | Trp           | W            |
| Tyrosine      | Tyr           | Y            |



There are *a lot* of different proteins out there, databases like [UniProt](http://www.uniprot.org/proteomes/UP000005640) have catalogued 70,940 different known proteins for humans alone.
The reason for this great diversity is twofold; there are 20 different amino acids in regular use, and each amino acid added to a peptide chain multiplies the possible combinations by 20.
What this means is that for every protein with a function, there are vastly more sequences that do nothing.

Let's look at insulin as an example. Human insulin, a "small" protein of 51 amino acids, aids in the control of blood glucose levels.
It is released by β-cells in the pancreas in response to glucose. 
If we were to generate all the other possible amino acids of this length, we'd have 20⁵¹ - 1(for insulin) different polypeptides.


That's a lot. There are only an estimated 10⁵⁰ [[1]](http://www.fnal.gov/pub/science/inquiring/questions/1atoms.html) atoms on earth. I don't think you need me to tell you that computing all of those would be extremely difficult. Sequences 7 or smaller, however, are completely do-able.

-----

### The Combinatorics Part


There are several different enumerative methods, or ways of arranging items.

**_Some of these are:_**

1. Permutations--order matters, you can't reuse objects.
 - Example: Possible ways of doing my morning routine; brush teeth, then shower, then eat? or eat, then brush, then shower?

2. Combinations--order doesn't matter, you can't reuse objects.
 - Example: I need some pencils for the exam tomorrow, which should I throw in my bag?

3. Sequences--order matters, you can reuse objects.
 - Example: My bank PIN, no I'm not gonna link that [here](https://www.youtube.com/watch?v=dQw4w9WgXcQ)


>I guess a _combination_ lock is really a _sequence_ lock

<img src="{{ site.baseurl }}/images/Combination_lock.jpg" alt="a sequence lock" style="width: 500px;"/>

So DNA, RNA, and proteins are all sequences, who would have thought...

In protein-ese "A-C-D-E" is different from "E-D-C-A" because we understand that by convention the left-to-right sequence goes from N-terminus to C-terminus.
We can also reuse letters to make sequences like "Y-Y-Y-Y-Y"


### Getting to the Code

This sounds like an iterator problem, luckily the standard library of Python comes with [itertools](https://docs.python.org/3/library/itertools.html). This module implements many different iterators in C, which will allow the program to execute a great deal faster than if it had been coded with a bunch of nested for loops in python.

To find all of the sequences for a given sequence length, we'll need to perform the Cartesian product.
Itertools has this built-in with the [product()](https://docs.python.org/3/library/itertools.html#itertools.product) method for generating sequences.

We can begin by importing our library, defining our amino acids, and the desired sequence length.
4-long should execute quick on most hardware and still prove to be a useful example

```python
from itertools import product

amino_acids = 'ACDEFGHIKLMNPQRSTVWY'
sequence_length = 4
```

Now we have to bring in the itertools.product() method using our defined parameters from above.
Since we are looping through all of the possible sequences, we should use a for loop.
Bear with me, I'll explain.

```python
for sequence in product(amino_acids, repeat=sequence_length):
  sequence_string = ''.join(sequence)
  print(sequence_string)
```

`itertools.product()` returns tuples for the cartesian product.
_i.e. `list(product('AB', 'AB'))` would be `[('A', 'A'), ('A', 'B'), ('B', 'A'), ('B', 'B')]`_
Alternatively, we can state the repeat which specifies the length of the individual tuples.

`itertools.product(sequence)` would default to a tuple length of 1
_i.e. (A,) (C,) ... (Y,)_
Not helpful.

Alternatively, adding, `repeat=` allows us to define the length of sequence generated.

`''.join(sequence)` takes the tuples and smushes them down into strings, this is useful because we'd rather think of our amino acids as a joined sequence. It also makes printing more readable and reduces the file size when we go to save these sequences later.
It works because `join()` takes the elements of the tuple and joins them by nothing, see [here](https://www.tutorialspoint.com/python/string_join.htm).


### Making it More Useful

The `print()` statement is great for spamming your console and all, but if we'd like to look back or analyze these sequences, we should save them to a file.

The easiest way to do this is a simple .txt file but theoretically you could store them in a csv, database, store as a bioinformatics sequence format like fasta, or anything else you can dream of. 

We can do this with the `with open('your_filename.txt') as file:` and `file.write(your_stuff_here)` statements.
Here's the full code with this tweak.

```python
from itertools import product

amino_acids = 'ACDEFGHIKLMNPQRSTVWY'
sequence_length = 4

with open('amino_acid_sequences.txt', 'a') as output_file:
    for sequence in product(amino_acids, repeat=sequence_length):
      sequence_string = ''.join(sequence)
      output_file.write(sequence_string + '\n')
```

Don't forget the newline `\n` character if you'd like to actually be able to read the text.

Be careful out there scientists, making the sequence length longer than 6 will start to enter gigabyte file size territory (increasing 20-fold each time) and runtimes will become ~20-fold longer for each additional peptide added.

